### PR TITLE
Test: tests for empty fields on Login form

### LINKED ISSUE #32: Test: write tests for input validations on Login form
## Description
As a developer,
I need check if a warning comes up for each of the form fields if input inserted is invalid,
so that I can make sure user insert a valid input.

### Update [Required]
- [ ] Write tests using React Testing Library in the Login.test.js file
- [ ] Test if validation message shows up for username
- [ ] Test if validation message shows up for password

### Enhancement to Update [Optional]
NA

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
3 hours

### PR DESCRIPTION
**Pull request by khushishikhu**
Description

testing for if the Username and Password field is empty

Fixes #32 

Type of Change:

- Code
- Quality Assurance
- Documentation

Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by khushishikhu** [comment]
@mtreacy002  i tried to make tests can you please review it

**Comment by mtreacy002** [comment]
Will check it out and let you know, @khushishikhu . But, can you change the title of the PR to `Test: ...` instead of `docs: ...` because this PR is for testing the UI 😉.

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: @khushishikhu . I thought we want to test if the input is not given (aka empty) a warning should be shown on the empty field as per what we've discussed [here](https://github.com/anitab-org/bridge-in-tech-web/issues/32#issuecomment-683351426). However the content shown here is for when the correct credentials are given the error will not be shown. We have that test already at the very top of the file ("allows the user to login successfully"). Can you please change it to checking if fields is/are empty?

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: If this is for checking empty field, the target value should be "")

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: the same here, but you can also leave it as it is as the warning will be shown on the first empty field (so, if you test when username is empty, that'll suffice for the test purposes).

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: Here we want it to show warning on that field like the screenshot below
Screen Shot 2020-09-01 at 11 44 59 am

File: src/tests/Login.test.jsx
Comment (mtreacy002) [reply src/tests/Login.test.jsx]: also, please be clear on the naming of the test. Naming it `input field` doesn't actually show what the purpose of the test. 😉

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@khushishikhu . I have made some suggestions here. Please check it out 😉

**Comment by khushishikhu** [comment]
thanks @mtreacy002  for letting me know.

File: src/tests/Login.test.jsx
Comment (khushishikhu) [reply src/tests/Login.test.jsx]: i tried to do that but then the tests are not passing and showing me the error

Comment (mtreacy002) [reply src/tests/Login.test.jsx]: @khushishikhu, perhaps you should use the [.toBeRequired()](https://github.com/testing-library/jest-dom#toberequired) check of @tesing-library/jest-dom instead?

**Comment by khushishikhu** [comment]
bit
@mtreacy002  i did all the required changes but i don't know why i am still getting these fail test please help me out

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: @khushishikhu. from the @testing-library/jest-dom .toBeRequired() I referred to you [here](https://github.com/anitab-org/bridge-in-tech-web/pull/83#discussion_r480874784) seems to be used with getByTestId. Have you tried giving a test id to the input field on the form and refer it on the test using `expect(getByTestId...` as shown on the jest-dom example?

**Review by mtreacy002 — COMMENTED** [review]
@khushishikhu, have you tried using getByTestId as per jest-dom doc example?

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: This is causing an error. This line is not needed because we are using .toBeRequired() to check the `required` attribute.

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: although not causing error, but this should come after `act`, so please move these 2 expect lines after the `act` block.

Comment (mtreacy002) [reply src/tests/Login.test.jsx]: Plus, .toBeRequired() doesn't take a parameter, so remove `""` from inside the parentheses `("")` -> `()`

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@khushishikhu , I have checked it out. There are few things causing the errors here.
1> line 131 (see comment)
2> line 122 (see comment)
3> line 16
Screen Shot 2020-09-02 at 3 10 19 pm
Now that we test for `required` which warning comes before sending the API request, you can remove line 16 (expcext(req.body).toEqual(payload)) and simply replace const payload to req.body, just as I've done here.
Screen Shot 2020-09-02 at 3 24 19 pm

Note: upon checking, I also see another way to get the html element instead of using "getByTestId" (nothing wrong with it, but to keep consistency across all other test cases, it's probably best to use the "getByLabelText" as how it was done in the `PersonalDetails.test.js` line 51++

Screen Shot 2020-09-02 at 3 36 44 pm

Noticed that in `PersonalDetails.jsx`, the `label` is used with `id` instead of `htmlFor` and `input` has an `aria-labelledBy` attribute to make the referencing works.

You can skip this (replacing getByTestId) if you're not sure how to do it. But if you could fix point 1 to 3 mentioned above, that'll be great. Let me know if you still need help.

**Comment by khushishikhu** [comment]
@mtreacy002  Thank you so much for guiding me. i have done all the changes you asked to do. the test is passing. please review it

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: expect is an assert, so not part of `act` so put it below line 125 `act` block (not inside the block itself). You can read more about 3As (Arrange, Act, and Assert) [here](https://testing-library.com/docs/react-testing-library/example-intro).
Also, the test still failed on Password field. Can you fix this?
Screen Shot 2020-09-02 at 6 20 42 pm

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@khushishikhu . You're pretty close to finish. This test still faild. Please check my feedback below.

File: src/tests/Login.test.jsx
Comment (mtreacy002) [reply src/tests/Login.test.jsx]: Don't forget we want to check the `required` attribute, so please use .toBeRequired() instead of .toHaveValue("")

Comment (khushishikhu) [reply src/tests/Login.test.jsx]: yeah sure....i have done that now please check. thank you

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: Please take your time when modifying the code base on the feedback given, @khushishikhu . You missed the one I pointed out to you [here](https://github.com/anitab-org/bridge-in-tech-web/pull/83#discussion_r481889755} about taking out `except` out of the `act` block. It's a good practice to double check the changes you make before you push again to make sure that you've addressed each point of the feedback.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@khushishikhu, can you make sure you run the test to see if its passed before you make another push? This one is still failing when I run `npm run test`. It still showing the same error as the screenshot I placed [above](https://github.com/anitab-org/bridge-in-tech-web/pull/83#discussion_r481889755). Have you checked what's causing it?

File: src/tests/Login.test.jsx
Comment (khushishikhu) [reply src/tests/Login.test.jsx]: yeah pardon me. i didn't actually noticed that because my tests were passing. i will be more careful from next time

File: src/tests/Login.test.jsx
Comment (khushishikhu) [reply src/tests/Login.test.jsx]: i got the problem why that password field was failing

**Comment by khushishikhu** [comment]
@mtreacy002  please review it. i think now its okay

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@khushishikhu . You can see from Travis CI that the build is still failing. To check why ist's failing on Travis you can do the following steps:
* click on the `details`
Screen Shot 2020-09-03 at 4 02 57 pm

* click on `build` link to see thee build log
Screen Shot 2020-09-03 at 4 03 07 pm

* analyse the log to see where things has gone wrong
Screen Shot 2020-09-03 at 4 03 26 pm

From the log you could see that you haven't changed line 16 as I pointed out on point 3 [here](https://github.com/anitab-org/bridge-in-tech-web/pull/83#pullrequestreview-480482723). Can you do that, thanks? Hopefully it will fix the issue then.

**Comment by khushishikhu** [comment]
@mtreacy002  done with all the errors. Thank you so much

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: Remember to remove unsused lines.

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: Remove this too.

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: It's best to start the test name in its present tense instead of continuous form as the test using the format `it ......`, so using `it + checks ...` feels more appropriate then `it + checking ... 

Note that the test name doesn't start in capital letter either.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@khushishikhu . Well done that you've made the test passed. It just needs few minor changes. After you've done this, I'll approve. But then we need one more maintainers to approve this as well.

**Comment by khushishikhu** [comment]
@mtreacy002  done with all the required changes you asked for . Thank you

File: src/tests/Login.test.jsx
Comment (mtreacy002) [inline src/tests/Login.test.jsx]: No need to declare `render ...` as constant. To be consistent with the other tests (and within this test) just use `screen.getByLabelText` where needed as you did in lines 114-119.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
Sorry, @khushishikhu, I must've missed this one. But you're very very close to done. Only one comment here for improvement then I'll approve 👍

**Comment by khushishikhu** [comment]
@mtreacy002  thank you. i have done all the changes and the tests are passing

**Review by mtreacy002 — APPROVED** [review]
@khushishikhu . You've done a great job. Thank you for taking care of all the requests and for contributing to the project. It looks good to me now. I'm approving this. Cc @anitab-org/bridgeintech-maintainers , can someone also please review this as well? Thanks

**Review by foongminwong — APPROVED** [review]
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done

2. All possible responses were tested as below:

   * _Test: tests for empty fields on Login Form_  
      _Screenshot/gif/url_:  

     test-bit-backend-83

      _Expected Result_:  As a user, if either the username/ email or password field is left blank, then it will give out error/reminder `Please fill out this field`.
      _Actual Result_: Same as expected

3. OS Version: Windows 10

@khushishikhu 🎉

**Pull request merged**
